server/src/main/java/org/elasticsearch/indices/SystemIndexMappingUpdateService.java

```diff
@@ -92,11 +92,21 @@ public void clusterChanged(ClusterChangedEvent event) {
         }
 
         // if we're in a mixed-version cluster, exit
-        if (state.nodes().getMaxNodeVersion().after(state.nodes().getSmallestNonClientNodeVersion())) {
+        if (state.hasMixedSystemIndexVersions()) {
             logger.debug("Skipping system indices up-to-date check as cluster has mixed versions");
+            logger.trace(() -> "Min versions: " + state.getMinSystemIndexMappingVersions());
             return;
         }
 
+        assert state.nodes().getMaxNodeVersion().equals(state.nodes().getSmallestNonClientNodeVersion())
```

Comment:
This assertion can't get merged, because in the future there will be cases where nodes have different versions/transport versions/index versions but will still have the same system index versions. However, it's useful for running bwc tests.